### PR TITLE
Require pattern matching exhaustiveness in the LF spec.

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Changes the rules for case expressions to require exhaustiveness.
This involves moving out the "well-formed case alternative" rules
to a separate judgement, and introducing a "pattern match
exhaustiveness" judgement.

The symbol ⊲ used for pattern match exhaustiveness comes from
sheaf theory and formal topology, where "A ⊲ B₁, …, Bₙ" means
"A is covered by { B₁, …, Bₙ }", indicating an exhaustive
coverage of A by possibly overlapping conditions B₁, …, Bₙ.
For example, see [this nlab page](https://ncatlab.org/nlab/show/posite).

It seemed appropriate.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/spec/daml-lf-1.rst [hunk 1]
```diff
@@ -1327,6 +1283,120 @@ Then we define *well-formed expressions*. ::
...
+
+                                ┌──────────────────┐
+  Well-formed case alternatives │ Γ ⊢ τ // alt : σ │
+                                └──────────────────┘
+
+      'variant' T (α₁:k₁) … (αₙ:kₙ) ↦ … | Vᵢ : σᵢ | …  ∈  〚Ξ〛Mod
+      σᵢ  ↠  σᵢ'
+      x : σᵢ'[α₁ ↦ τ₁, …, αₙ ↦ τₙ] · Γ  ⊢  e : σ
+    ——————————————————————————————————————————————————————————————— AltVariant
+      Γ  ⊢  Mod:T τ₁ … τₙ  //  Mod:T:V x  →  e : σ
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:1302]:
```suggestion
      Γ  ⊢  Mod:T τ₁ … τₙ  //  Mod:T:Vᵢ x  →  e : σ
```
or drop the `i` indices above, which probably makes it easier to read.

File: daml-lf/spec/daml-lf-1.rst [hunk 2]
```diff
@@ -1327,6 +1283,120 @@ Then we define *well-formed expressions*. ::
...
+    'variant' T (α₁:k₁) … (αᵣ:kᵣ) ↦ V₁ : σ₁ | … | Vₘ : σₘ  ∈  〚Ξ〛Mod
+    i₁, i₂, …, iₘ  ∈  {1, …, n}
+    altᵢ₁  =  Mod:T:V₁ x₁  →  e₁
+    altᵢ₂  =  Mod:T:V₂ x₂  →  e₂
+           ⋮
+    altᵢₘ  =  Mod:T:Vₘ xₘ  →  eₘ
+    ——————————————————————————————————————————————————————————————— ExhaustVariant
+    Mod:T τ₁ … τᵣ  ⊲  alt₁, …, altₙ
+
+    'enum' T … ↦ E₁ | … | Eₘ  ∈  〚Ξ〛Mod
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:1362]: Aren't enums always monomorphic?
```suggestion
    'enum' T ↦ E₁ | … | Eₘ  ∈  〚Ξ〛Mod
```

File: daml-lf/spec/daml-lf-1.rst [hunk 3]
```diff
@@ -1327,6 +1283,120 @@ Then we define *well-formed expressions*. ::
...
+    altᵢ₁  =  Mod:T:E₁  →  e₁
+    altᵢ₂  =  Mod:T:E₂  →  e₂
+           ⋮
+    altᵢₘ  =  Mod:T:Eₘ  →  eₘ
+    ——————————————————————————————————————————————————————————————— ExhaustEnum
+    Mod:T  ⊲  alt₁, …, altₙ
+
+    i₁, i₂  ∈  {1, …, n}
+    altᵢ₁  =  'Nil'  →  e₁
+    altᵢ₂  =  'Cons' xₕ xₜ  →  e₂
```
Comment (hurryabit) [inline daml-lf/spec/daml-lf-1.rst:1373]: Does our font have `j` (or `k`) as an index character? Then we could avoid one level of indexing.

**Review by hurryabit — APPROVED** [review]
I really like this formalization. Thank you so much!

File: daml-lf/spec/daml-lf-1.rst (hunk 3 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst:1373]: Yes, there's a bunch of subscripts. I'll use i,j for these 2-case ones.

File: daml-lf/spec/daml-lf-1.rst (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst:1302]: Good point!

File: daml-lf/spec/daml-lf-1.rst (hunk 2 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst:1362]: Oops, yes!

File: daml-lf/spec/daml-lf-1.rst (hunk 1 above)
Comment (sofiafaro-da) [reply daml-lf/spec/daml-lf-1.rst:1302]: I dropped the i indices and renamed σᵢ to τ and σᵢ' to τ', reserving σ only for the output type.

**Review by remyhaemmerle-da — APPROVED** [review]

**Pull request merged**